### DISCUSSION TITLE
Is static lambda really static?

### DISCUSSION THREAD

**Discussion by kefyru** [dotnet/runtime#69426] [General]
```cs
     Func<int,int> f = static i => i;
     Console.WriteLine(f.Method.IsStatic) // -> false !!!
```

Is it necessary to use 'static' keyword, or compiler can resolve it from lambda body?

**Comment by stephentoub** [comment]
The static keyword when applied to a lambda has zero impact on the generated code. It serves only to enforce that the body of the lambda doesn't close over non-static state. Whether the emmitted method for the lambda is actually an instance or static method is an implementation detail of the C# compiler, and as you saw, your code today will result in an instance method (for performance reasons).

**Reply by juwens** [reply]
Thx for explaining.

This is kinda weird/unexpected IMO. The keyword static should (for the sake of consistency) mark a static method (lambda or real) or class.

If the purpose is to only capture const variables, then the designers should have really chosen a different keyword.

I'm seeing a trend here, c# designers apparently prefer less keywords and therefore reuse existing keywords for new scenarios, which can be quite ambiguous/unintuitive like in lambdas. Introducing a new (and explicit) keyword would have been the alternative.

**Reply by tannergooding** [reply]
Whether it is actually static or not is largely an implementation detail.

Actually being static would be a de-optimization due to how delegates work (and needing to shuffle arguments around to go from the instance `Invoke` method on the `delegate` to the actual method being called if it were static).

The general intent here of `static` is to prevent implicit capturing of instance or local state, which is effectively how `static` methods work at the conceptual level. That is, you only get access to other static state and everything else has to be passed in.

**Reply by juwens** [reply]
yes, we understood that. The point is: the chosen keyword has a bad semantic for this new scenario.

Because: it does something completely different and counterintuitive, than the static keyword did in the last 22 years.

**Reply by davidfowl** [reply]
It's as if you declared the function as a static method which would be unable to any state other than what was passed to the method and other static variables. When you think about it that way, the keyword makes a lot of sense.

**Reply by tannergooding** [reply]
Yep!

I like to think of `static` as meaning it doesn't implicitly pull in the higher scope. Thus a `static` method in a class means it doesn't get the implicit instance scope. A `static` lambda/local method then effectively means the same, it doesn't get the implicit instance scope or the implicit local scope of the containing method.

Whether it is actually an instance method (that is the first parameter at the ABI level is `this`) or not is completely irrelevant to the visible semantics at the language level. This is particularly true for `lambdas` where you cannot do things like get a direct function pointer to them and so you can consider it an internal optimization detail of the compiler that will effectively never matter to the author of the code.

**Reply by juwens** [reply]
You again deliver an example, why it's quite poorly named. Static is for the last 22 years about object instances or the lack of it.

The meaning in the lambda is: "do not implicitly capture ANY local scoped variables", not capturing "this" is only ONE specialized case. The overlap with conventional static is therefore only a coincidence.

It's like calling a every car a police-car, despite most of the cars are no police cars.

I mean you can, for one scenario, change the meaning of a 22 year old keyword. But i wouldn't call it best practice.

```
static void Foobar()
{
    string foo = "";
    My.Lib.Func(static () => Console.Write(foo)); // compile error, though no this is involved
}
```

IMO a word like `stateless` or `nocapture` would suit the scenario better.

**Reply by tannergooding** [reply]
> IMO a word like stateless or nocapture would suit the scenario better.

This would have introduced a new contextual keyword, which has its own downsides and disadvantages. Many of which are repeatedly raised on the dotnet/csharplang repo amongst the many championed and proposed changes to the language.

> Static is for the last 22 years about object instances or the lack of it.

Sure, it meant that you didn't have implicit access to other instance members. Static lambdas also do not allow this access and so provide functionally the same behavior. They then provide an additional behavior of preventing all implicit captures. Conceptually, an instance member implicitly captures and tracks `this` and so its functionally very similar and a "natural extension" of the existing meaning.

Languages (spoken, written, and programming) evolve over time with meanings of words expanding or changing. The language designers carefully consider these evolutions of keywords and weigh the tradeoffs when making such decisions. This is also true of other languages, not just C# and you can see similar changes across a wide range of popular programming languages over time.

In any case, its far too late to change at this point and similar changes and minor evolutions of existing keywords will likely continue happening in the future where the C# language design team believes they make sense. Developers will need to learn and adapt to these changes, same as when encountering any new feature or when encountering a new word or a new usage of a word in a spoken/written language.

**Reply by CyrusNajmabadi** [reply]
I was the language designer here. Static was very intentionally used as it means do not capture 'this' (and the members implicitly available off of 'this'). Static on lambda serves the same purpose. Do not capture anything (explicitly or implicitly). 

Note, this also exactly matches the same semantics with static local functions. Picking a different modifier to mean the same thing would have introduced a very undesirable inconsistency

**Reply by CyrusNajmabadi** [reply]
> This is kinda weird/unexpected IMO. The keyword static should (for the sake of consistency) mark a static method (lambda or real) or class.

It does.  It makes the lambda conceptually static. Importantly, it enforces that aspect so you can't accidentally screw it up later. 

Now, it sounds like you feel that should have a specific impact on emitting.   We disagree. The language has firmly held the belief that how it emits code is an implementation detail.  Lambdas (and local functions) only exist in method bodies. And as such we consider how we emit entirely an implementation detail that we can and do change depending on the benefits or drawbacks we get from it.

**Reply by juwens** [reply]
Thanks for the detailed explanation.
I still think it wasn't the best idea.

**Reply by Xyncgas** [reply]
When you apply static it constrains the lambda to capture only static stuff, if it's not static it enables you to capture the current lexical context, of course instructions are static.

Compiler can, hey I didn't make roslyn but, it can remember the lexical context of the instance and probably store it somewhere so you can pass the lambda around and it still can use those references else where, so it would create a storage for every lambda that's in a class and it would do it again if you make another instance of the class, compiler can optimize these creation, but compared to static lambda it's just created once and it's there for you to pass around and use it.

Of course in theory you can have a lambda in a class and have a static instance of that class and the lambda lives forever with the object and that's basically like creating a static lambda

A static λ(f -> λx·2 * x) is also just ```int f (x :int) => 2 * x;``` in C#

In another way lambda in c# is an object you can pass around that has a static function which is the code you wrote in the lambda and references which is the stuffs you are able to use or optimally the stuffs you used in those codes so when you call the lambda these stuffs are automatically applied

**Reply by Xyncgas** [reply]
meanwhile for interop ability with F# curried functions, it's static lambda in C# though it's annoying to use ```MyFSharpModule.MyFunction.Invoke(x).Invoke(y).Invoke()```
Here MyFunction is under MyFSharpModule ```module``` in F#, if compiled to C# it would be under a static class called MyFSharpModule

**Reply by En3Tho** [reply]
Don't forget that introducing a new keyword to the language should hold more value than "possibly a better description for what's happening". It's a very good practice to keep the same keyword even for things that might  not be exactly the same in different context but still very similar.

In this context I know how static behaves for usual methods, how I can only use direct parameters, const values or other statics. That's it. It's the same with static lambdas: direct lambda parameters, consts or other statics. Yes lambdas are generated differently and it is not exactly the same but the main idea is very similar.

Imagine we add a different keyword here, whatever it might be. Not only everyone has to learn it now, you also have to carefully explain why:

1) it's such a big difference that it deserves a whole new keyword
2) why "static" was not a good choice
3) why new keyword only applies to lambdas a not methods
4) if it applies only a very specific case and means the very similar but "very different" thing then what's the point of it?

You also shouldn't forget that C# already has a LOT to learn keywords included. Adding another one for very little value (if any) will just complicate things for everybody, especially for newcomers.

**Reply by juwens** [reply]
Cool, so now as static isn't about instance anymore and only about scope, we can use static classes as generic parameter like in `CreateLogger<MyStaticClass>()`, right?

**Reply by juwens** [reply]
How about a keyword which does not allow a method or lambda to access any static or instance variables in it's class a method can only access parameters, how would we call that?

Following your logic it would be `static static`, because we must not introduce a new keyword. And because we can imagine the first static as the classic static (do not allow access to instance fields/this), and the second static for do not allow access to anything out of my current scope (the method/lambda).

If you think about that, you might understand why misusing the existing keyword was a poor choice.

**Reply by juwens** [reply]
`Adding another one for very little value (if any) will just complicate things for everybody, especially for newcomers` you think that one keyword which now does different things (1. instance capturing, 2. inconsistent variable capturing) is less confusing for newcomers?

**Reply by CyrusNajmabadi** [reply]
> Cool, so now as static isn't about instance anymore and only about scope, we can use static classes as generic parameter 

Yes. This is something we're discussing. Especially since it would make sense for a static class to implement an interface with just static abstract methods.

**Reply by juwens** [reply]
The docs are a witness of this ambiguity too.

The first sentence is: `Use the static modifier to declare a static member, which belongs to the type itself rather than to a specific object.` It clearly states `a static member`.

Which is later extended/modified to usecases like local functions and lambdas. `A static lambda or anonymous method can't capture local variables or instance state.`

https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/static

**Reply by CyrusNajmabadi** [reply]
> How about a keyword which does not allow a method or lambda to access any static or instance variables in it's class a method can only access parameters, how would we call that?

We don't see any value in that. There was value in not capturing instance state. 

> Following your logic it would be static static, 

No. There's no logic that leads to that. That's a figurative strawman. You've invented a non-existent proposal to them try to knock down. 

> because we must not introduce a new keyword. 

Introducing a new keyword is fine if it's the best option. Here it is not.  Indeed, it would be actively problematic for the inconsistency with the rest of the language (especially static local functions). It's very important to us that lambdas and local functions naturally tie together.  Using another keyword here would directly harm that goal. 

> And because we can imagine the first static as the classic static (do not allow access to instance fields/this), 

Your strawman is non-sensical. We literally are using static for it's traditional meaning. As you *yourself* point out, it's to prevent accessing instance state. That's been the "classic" meaning, and it's the *same* meaning we want for local functions *and* for lambdas. It's literally being consistent with the language, as you yourself point out. Your strawman makes no sense because:

1. We have no interest in a Keyword to prevent accessing statics.  
2. If we had that interest, we wouldn't pick 'static' for that purpose because that would make no Sense. 


> If you think about that, you might understand why misusing the existing keyword was a poor choice.

Your argument makes no sense. It complains about us using the existing keyword with three same semantics for how it already works. Then it says we made a mistake because that keyword would *not* make sense for a hypothetical use case where we would not use that keyword *anyways*.

**Reply by CyrusNajmabadi** [reply]
> you think that one keyword which now does different things 

It doesn't do different things. It's the same in both cases. You cannot access instance state. That's literally what 'static' means. We use it for the same meaning in both cases.  For both, any attempt to access 'this' will fail.

**Reply by CyrusNajmabadi** [reply]
> The docs are a witness of this ambiguity too.

The docs say the exact same thing.  A lambda/local function is not a member in the first place. So there's no concept of there being a type to access them off of. So the 'access off of type vs instance' part simply does not apply. 

But for both, the 'static' modifier means `cannot access 'this'`.

**Reply by agocke** [reply]
As the author of closure conversion, I feel compelled to remind everyone: using reflection on delegates produced by lambdas (i.e., retrieving the `MethodInfo` via `.Method`) is explicitly compiler-defined behavior and you are not guaranteed any specific results or stability. Depending on this behavior is prohibited by the C# language specification.

**Reply by TonyValenti** [reply]
@agocke - this is likely more of a compiler or GUI request but is there any way that you can surface everything captured by a closure?  We've had many situations where a closure results in a long living invisible reference to a giant object we want GCed.

**Reply by OwnageIsMagic** [reply]
@TonyValenti see https://github.com/controlflow/resharper-heapview and https://github.com/microsoft/RoslynClrHeapAllocationAnalyzer

**Reply by colejohnson66** [reply]
> As the author of closure conversion, I feel compelled to remind everyone: using reflection on delegates produced by lambdas (i.e., retrieving the `MethodInfo` via `.Method`) is explicitly compiler-defined behavior and you are not guaranteed any specific results or stability. Depending on this behavior is prohibited by the C# language specification.

What's the logic behind attaching an "instance" to `static` lambdas? I've seen it mentioned a few times that it's for performance reasons, but I haven't found a good explanation of _why_.

**Reply by colejohnson66** [reply]
> > Cool, so now as static isn't about instance anymore and only about scope, we can use static classes as generic parameter
> 
> Yes. This is something we're discussing. Especially since it would make sense for a static class to implement an interface with just static abstract methods.

Is there any update on this? I actually ran into this problem over the weekend trying to do something like `public static class AddCarry : IGenericAluOperations` in an emulator, but was told that static classes can't have "extends" lists (seems to be Rider's own message for CS0714). Removing `static` made Rider complain about a non-instantiated class.

**Reply by OwnageIsMagic** [reply]
@colejohnson66 https://stackoverflow.com/a/77004189/5647513

**Reply by colejohnson66** [reply]
@OwnageIsMagic Again, I understand it's for performance reasons, but I've yet to see _why_. "Performance" is not an answer.

**Reply by tannergooding** [reply]
`delegates` are types that wrap a function pointer and may track an associated instance. To invoke a delegate, you literally call an instance method named `Invoke(...)`

Since `Invoke(...)` is (and must be) an instance method so that it can access the function pointer and tracked state, this means there is a hidden first `this` parameter.

Thus, for a delegate bound to an instance method, it can simply replace the `this` parameter that's already been passed with the underlying tracked instance the delegate was bound against.

While for a static method, it means all the arguments are positionally "off by one" and need to be shuffled around (often swapping registers, potentially spilling or copying memory, etc). This "fixup" isn't exactly expensive, but it isn't free either.

Thus, using instance methods is functionally invisible to the user (it's an implementation detail of the emitted IL) and avoids the performance penalty that fixing up the signature to match ABI requirements would otherwise cause.
